zonemaster/zonemaster-engine
# Caching: add global cache based on Redis (experimental)

## Purpose

Improve our cache implementation to make room to integrate other caching solutions. Provide an optional and experimental global cache layer based on Redis. This can be handy when multiple Zonemaster instances run concurrently (when dealing with batches for instance).

## Context

https://gitlab.rd.nic.fr/zonemaster/zonemaster-engine/-/commit/b96181e460b5824dcd4ac913a3210c458970126a
and
https://gitlab.rd.nic.fr/zonemaster/zonemaster-engine/-/commit/0073398f8a4c47bb6f7f7d8dae8f8a2ddd2694a6

Credits to @blacksponge for the initial proposal

## Changes

* make Cache.pm an interface
* move old Cache.pm logic into a LocalCache module
* new RedisCache module (loaded only if configured in profile.json).

## How to test this PR

1. install Redis and the following perl modules: Redis and Data::MessagePack. For Debian, this can be done with:
     ```
     apt install redis libredis-perl libdata-messagepack-perl
     ```
2. start Redis service
3. modify the profile.json file to specify the Redis server (see `share/profile_additional_properties.json`) :
     ```
     ...
    "cache": {
       "redis": {
          "server": "127.0.0.1:6379",
          "expire": 60
        }
     },
    ...
    ```
4. run the tests or run Zonemaster on a zone to test
5. check that there is data stored in Redis
    ```
    redis-cli KEYS 'ns*'
    ```
